File: TheGarageMvc/TheGarageMvc/Hubs/deviceCommunicationHub.js
$(function () {
    
    var lightState = ["off", "off"];
    var softlockState = false;
    var doorState = ["down", "down"];

    $.connection.deviceCommunicationHub.client.onDoorChange = function (index, status) {
        var img = 'Images/unknown.png';

        if (status === 'down') {
            img = 'Images/closed.png';
        } else if (status === 'up') {
            img = 'Images/open.png';
        }

        $('#door' + index).attr('src', img);

        doorState[index] = status;
    };

    $.connection.deviceCommunicationHub.client.onLightChange = function (index, status) {
        var img = 'Images/lighton.png';
        
        if (status === "off") {
            img = 'Images/lightoff.png';
        }
        
        $('#light' + index).attr('src', img);
        
        lightState[index] = status;
    };

    $.connection.deviceCommunicationHub.client.onLockChange = function (name, garageLocked, hardlock, softlock) {

        var img = 'Images/unlocked.png';
        
        if (softlock) {
            img = 'Images/locked.png';
        }

        $('#softlockImage').attr('src', img);
        $('#garageName').text(name);
        $('#garageLockStatus').text(garageLocked ? "Locked" : "Unlocked");
        $('#hardlockStatus').text(hardlock ? "Locked" : "Unlocked");

        softlockState = softlock;
    };


    $('.lights').click(function () {
        var name = $(this).attr('id');
        var index = +name.substr(name.length - 1, 1);

        var cmd = "on";

        if (lightState[index] === "on") {
            cmd = "off";
        }
        
        $.connection.deviceCommunicationHub.server.activateLight(index, cmd);
    });

    $('#softlockImage').click(function () {
        var cmd = 'lock';
        
        if (softlockState) {
            cmd = 'unlock';
        }
        
        $.connection.deviceCommunicationHub.server.activateSoftLock(cmd);
    });

    $('.doors').click(function () {
        var name = $(this).attr('id');
        var index = +name.substr(name.length - 1, 1);

        var cmd = 'close';
        if (doorState[index] === "down") {
            cmd = 'open';
        } else if (doorState[index] === "unknown") {
            cmd = 'toggle';
        }
        
        $.connection.deviceCommunicationHub.server.activateDoor(index, cmd);
    });

    $.connection.hub.logging = true;

    $.connection.hub.start().done(function() {
        $.connection.deviceCommunicationHub.server.joinGroup('WebApp');
    });

});

File: TheGarageMvc/TheGarageMvc/Hubs/DeviceCommunicationHub.cs
using Microsoft.AspNet.SignalR;
using Newtonsoft.Json;
using System;
using System.Configuration;
using System.Threading.Tasks;
using TheGarageMvc.HubModels;
using TheGarageMvc.PushNotifications;

namespace TheGarageMvc.Hubs
{
    public class DeviceCommunicationHub : Hub
    {
        public byte[] Key { get; set; }
        public byte[] InitializationVector { get; set; }
        public string Password { get; set; }
        public static string ConsoleKey { get; set; }

        readonly NotificationBroker _notificationBroker = new NotificationBroker();

        public DeviceCommunicationHub()
        {
            Key = Convert.FromBase64String(ConfigurationManager.AppSettings["SharedSecret"]);
            InitializationVector = Convert.FromBase64String(ConfigurationManager.AppSettings["SharedIV"]);
            Password = ConfigurationManager.AppSettings["Password"];
        }

        public FromServerToClientData Request(FromClientToServerData request)
        {
            var response = new FromServerToClientData { Text = "Responding to: " + request.Text };

            if (request.Text == "SendMessage")
            {
                var sendResponse = new FromServerToClientData { Text = "WooHoo!" };

                Clients.Others.OthersCallback(sendResponse);
            }

            if (request.Text == "SendConsole")
            {
                var consoleResponse = new FromServerToClientData { Text = "Calling Mr. Console!", Command = request.Command };

                Clients.Client(ConsoleKey).ActivateDoor(0, "on");

                Clients.Client(ConsoleKey).OthersCallback(consoleResponse);
            }

            return response;
        }

        public void OnDoorChange(int id, string command)
        {
            Clients.All.OnDoorChange(id, command);

            _notificationBroker.Send("Garage Door", command);
        }

        public void OnLightChange(int id, string command)
        {
            Clients.All.OnLightChange(id, command);

            switch (id)
            {
                case 0:
                    _notificationBroker.Send("Inside Lights", command);
                    break;

                case 1:
                    _notificationBroker.Send("Outside Lights", command);
                    break;
            }
        }

        public void OnLockChange(string name, bool garage, bool hardlock, bool softlock)
        {
            Clients.All.OnLockChange(name, garage, hardlock, softlock);

            if ( hardlock || softlock ) 
                _notificationBroker.Send("Garage", "Locked");
            else 
                _notificationBroker.Send("Garage", "UnLocked");
        }

        public void ActivateDoor(int id, string command)
        {
            Clients.Client(ConsoleKey).ActivateDoor(id, command);
        }

        public void ActivateLight(int id, string command)
        {
            Clients.Client(ConsoleKey).ActivateLight(id, command);
        }

        public void ActivateSoftLock(string command)
        {
            Clients.Client(ConsoleKey).ActivateSoftLock(command);
        }

        public override Task OnConnected()
        {
            Clients.Client(ConsoleKey).RequestStatus(Context.ConnectionId);
            return base.OnConnected();
        }

        public Response SignOn(SignOnRequest signOnRequest)
        {
            var response = new Response { Version = 1 };

            var identity = DeviceHubIdentity.Decrypt(signOnRequest.Identity, Key, InitializationVector);

            response.Payload = JsonConvert.SerializeObject(identity.Credential.Password == Password ? "Success" : "Unauthorized");

            ConsoleKey = identity.ConnectionId;

            return response;
        }

        public void JoinGroup(string groupName)
        {
            Groups.Add(Context.ConnectionId, groupName);
        }
    }

}